File: services/gui/src/app/controllers/title.js
var Title = require("../models/title"),
    mongoose = require("mongoose"),
    HttpStatus = require("http-status-codes")

const superagent = require('superagent')

//CREATE POST /title to add a new title
function postTitle(req, res){
    //metrics, but not during test
    if(process.env.NODE_ENV !== "test"){
        console.log("metrics.postTitle")
    }

    var newTitle = new Title({...req.body.title, user: req.user});

    //TODO: use dynamic name and port for ketchup-service
    //TODO: switch to https (self-signed)
    //TODO: check tomatoURL upfront, if empty skip ketchup request
    //get path of tomatoURL
    var path
    if(req.body.title.tomatoURL){
        path = req.body.title.tomatoURL.substring(req.body.title.tomatoURL.indexOf("m/")+4)
    }

    superagent.get(`http://${process.env.KETCHUP_ENDPOINT}/${path}`)
        .end((err, response) => {
            if (err) {
                console.log(`WAR: 🍅 KETCHUP failed us 😭, assuming there is no rating, here is the reason: ${err}`)
                newTitle.tomatoUserRating = -1
            } else {
                newTitle.tomatoUserRating = response.body.tomatoUserRating
            }

            newTitle.save((err, title) => {
                if(err) {
                    res.status(HttpStatus.NOT_FOUND)
                      .send(err)
                } else {
                    //respond with JSON when asked (for API calls and integration testing), otherwise render HTML
                    if(req.get('Accept') === "application/json"){
                        res.status(HttpStatus.CREATED)
                          .json({message: "Title successfully added!", title})
                    } else {
                        req.flash("success", "You've added '" + title.name + "'' to your watchlist!")
                        res.redirect("title")
                    }
                }
            })
        }
    );
}

//READ GET /title to retrieve all titles
function getTitles(req, res){
    //metrics, but not during test
    if(process.env.NODE_ENV !== "test"){
        console.log("metrics.getTitles")
    }
    var query = Title.find({user:req.user}, undefined, {sort:{createdAt:-1}});
    query.exec((err, titles) => {
        if(err) {
            res.status(HttpStatus.NOT_FOUND)
                .send(err)
        } else {
            //respond with JSON when asked (for API calls and integration testing), otherwise render HTML
            if(req.get('Accept') === "application/json"){
                res.json(titles)
            } else {
                res.render("title/index", {titles: titles})
            }
        }
    })
}

//READ GET /title/:id to retrieve a title
function getTitle(req, res){
    //metrics, but not during test
    if(process.env.NODE_ENV !== "test"){
        console.log("metrics.getTitle")
    }
    var query = Title.findOne({ _id: req.params.id, user: req.user });
    query.exec((err, title) => {
        if(err || !title) {
            res.status(HttpStatus.NOT_FOUND)
                .send(err)
        } else {
            //respond with JSON when asked (for API calls and integration testing), otherwise render HTML
            if(req.get('Accept') === "application/json"){
                res.json(title)
            } else {
                res.redirect("/title")
            }
        }
    })
}

//UPDATE PUT /title/:id to update a title
function updateTitle(req, res){
    //metrics, but not during test
    if(process.env.NODE_ENV !== "test"){
        console.log("metrics.updateTitle")
    }
    //check for name in body
    if(req.body.title.name !== "") {
        var query = Title.findOneAndUpdate({ _id: req.params.id, user: req.user }, {...req.body.title, user: req.user}, {new: true})
        query.exec((err, updatedTitle) => {
            if(err || !updatedTitle) {
                res.status(HttpStatus.NOT_FOUND)
                    .send(err)
            } else {
                    //respond with JSON when asked (for API calls and integration testing), otherwise render HTML
                    if(req.get('Accept') === "application/json"){
                        res.json({message: "Title successfully updated!", updatedTitle})
                    } else {
                        if(req.body.title.seen) {
                            if(updatedTitle.seen === true)
                            {
                                req.flash("success", "You've marked '" + updatedTitle.name + "' as seen!")
                            } else {
                                req.flash("success", "You've marked '" + updatedTitle.name + "' as unseen!")
                            }
                        } else {
                            req.flash("success", "You've changed the name to '" + updatedTitle.name + "'!")
                        }
                        res.redirect("/title")
                    }
            }
        })
    } else {
        //respond with JSON when asked (for API calls and integration testing), otherwise render HTML
        if(req.get('Accept') === "application/json"){
            res.status(HttpStatus.BAD_REQUEST)
                .send("You need to specify a name and it can't be empty!")
        } else {
            res.redirect("/title")
        }
    }
}

//DELETE DELETE /title/:id to delete a title
function deleteTitle(req, res){
    //metrics, but not during test
    if(process.env.NODE_ENV !== "test"){
        console.log("metrics.deleteTitle")
    }
    var query = Title.findOneAndRemove({ _id: req.params.id, user: req.user }, req.body.title)
    query.exec((err, deletedTitle) => {
        if(err || !deletedTitle) {
            res.status(HttpStatus.NOT_FOUND)
                .send(err)
        } else {
            //respond with JSON when asked (for API calls and integration testing), otherwise render HTML
            if(req.get('Accept') === "application/json"){
                res.json({message: "Title successfully deleted!", deletedTitle})
            } else {
                req.flash("success", "You've deleted '" + deletedTitle.name + "'' from your watchlist!")
                res.redirect("/title")
            } 
        }
    })
}

//export all functions
module.exports = { getTitles, getTitle, postTitle, updateTitle, deleteTitle };

File: services/gui/app/controllers/title.go
package controllers

import (
	"database/sql"
	"encoding/json"
	"net/http"
	"time"

	"github.com/gorilla/mux"
	"github.com/gorilla/sessions"
	"github.com/sirupsen/logrus"
	"gopkg.in/mgo.v2/bson"
)

var (
	db     *sql.DB
	store  *sessions.CookieStore
	logger = logrus.New()
)

type Title struct {
	ID              bson.ObjectId `bson:"_id,omitempty"`
	Name            string        `json:"name"`
	CreatedAt       time.Time     `json:"createdAt"`
	Seen            bool          `json:"seen"`
	SeenOn          time.Time     `json:"seenOn"`
	Poster          string        `json:"poster"`
	ImdbRating      float64       `json:"imdbRating"`
	ImdbID          string        `json:"imdbID"`
	Year            string        `json:"year"`
	TomatoUserRating float64      `json:"tomatoUserRating"`
	TomatoURL       string        `json:"tomatoURL"`
	User            string        `json:"user"`
	Genres          []string      `json:"genres"`
}

func postTitle(w http.ResponseWriter, r *http.Request) {
	session, _ := store.Get(r, "session-name")
	userID := session.Values["user_id"].(string)

	var newTitle Title
	err := json.NewDecoder(r.Body).Decode(&newTitle)
	if err != nil {
		http.Error(w, err.Error(), http.StatusBadRequest)
		return
	}

	newTitle.User = userID
	newTitle.CreatedAt = time.Now()

	if newTitle.ImdbRating == 0 {
		newTitle.ImdbRating = -1
	}

	if newTitle.TomatoURL != "" {
		path := newTitle.TomatoURL[strings.Index(newTitle.TomatoURL, "m/")+2:]
		resp, err := http.Get("http://" + os.Getenv("KETCHUP_ENDPOINT") + "/" + path)
		if err != nil {
			logger.Warnf("KETCHUP failed: %v", err)
			newTitle.TomatoUserRating = -1
		} else {
			defer resp.Body.Close()
			var result map[string]interface{}
			json.NewDecoder(resp.Body).Decode(&result)
			newTitle.TomatoUserRating = result["tomatoUserRating"].(float64)
		}
	}

	query := `INSERT INTO titles (name, createdAt, seen, seenOn, poster, imdbRating, imdbID, year, tomatoUserRating, tomatoURL, user, genres) 
			  VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)`
	_, err = db.Exec(query, newTitle.Name, newTitle.CreatedAt, newTitle.Seen, newTitle.SeenOn, newTitle.Poster, newTitle.ImdbRating, newTitle.ImdbID, newTitle.Year, newTitle.TomatoUserRating, newTitle.TomatoURL, newTitle.User, pq.Array(newTitle.Genres))
	if err != nil {
		http.Error(w, err.Error(), http.StatusInternalServerError)
		return
	}

	w.WriteHeader(http.StatusCreated)
	json.NewEncoder(w).Encode(newTitle)
}

func getTitles(w http.ResponseWriter, r *http.Request) {
	session, _ := store.Get(r, "session-name")
	userID := session.Values["user_id"].(string)

	query := `SELECT id, name, createdAt, seen, seenOn, poster, imdbRating, imdbID, year, tomatoUserRating, tomatoURL, user, genres 
			  FROM titles WHERE user = $1 ORDER BY createdAt DESC`
	rows, err := db.Query(query, userID)
	if err != nil {
		http.Error(w, err.Error(), http.StatusInternalServerError)
		return
	}
	defer rows.Close()

	var titles []Title
	for rows.Next() {
		var title Title
		err := rows.Scan(&title.ID, &title.Name, &title.CreatedAt, &title.Seen, &title.SeenOn, &title.Poster, &title.ImdbRating, &title.ImdbID, &title.Year, &title.TomatoUserRating, &title.TomatoURL, &title.User, pq.Array(&title.Genres))
		if err != nil {
			http.Error(w, err.Error(), http.StatusInternalServerError)
			return
		}
		titles = append(titles, title)
	}

	w.Header().Set("Content-Type", "application/json")
	json.NewEncoder(w).Encode(titles)
}

func getTitle(w http.ResponseWriter, r *http.Request) {
	session, _ := store.Get(r, "session-name")
	userID := session.Values["user_id"].(string)

	vars := mux.Vars(r)
	titleID := vars["id"]

	query := `SELECT id, name, createdAt, seen, seenOn, poster, imdbRating, imdbID, year, tomatoUserRating, tomatoURL, user, genres 
			  FROM titles WHERE id = $1 AND user = $2`
	row := db.QueryRow(query, titleID, userID)

	var title Title
	err := row.Scan(&title.ID, &title.Name, &title.CreatedAt, &title.Seen, &title.SeenOn, &title.Poster, &title.ImdbRating, &title.ImdbID, &title.Year, &title.TomatoUserRating, &title.TomatoURL, &title.User, pq.Array(&title.Genres))
	if err != nil {
		http.Error(w, err.Error(), http.StatusNotFound)
		return
	}

	w.Header().Set("Content-Type", "application/json")
	json.NewEncoder(w).Encode(title)
}

func updateTitle(w http.ResponseWriter, r *http.Request) {
	session, _ := store.Get(r, "session-name")
	userID := session.Values["user_id"].(string)

	vars := mux.Vars(r)
	titleID := vars["id"]

	var updatedTitle Title
	err := json.NewDecoder(r.Body).Decode(&updatedTitle)
	if err != nil {
		http.Error(w, err.Error(), http.StatusBadRequest)
		return
	}

	query := `UPDATE titles SET name = $1, seen = $2, seenOn = $3, poster = $4, imdbRating = $5, imdbID = $6, year = $7, tomatoUserRating = $8, tomatoURL = $9, genres = $10 
			  WHERE id = $11 AND user = $12 RETURNING id, name, createdAt, seen, seenOn, poster, imdbRating, imdbID, year, tomatoUserRating, tomatoURL, user, genres`
	row := db.QueryRow(query, updatedTitle.Name, updatedTitle.Seen, updatedTitle.SeenOn, updatedTitle.Poster, updatedTitle.ImdbRating, updatedTitle.ImdbID, updatedTitle.Year, updatedTitle.TomatoUserRating, updatedTitle.TomatoURL, pq.Array(updatedTitle.Genres), titleID, userID)

	var title Title
	err = row.Scan(&title.ID, &title.Name, &title.CreatedAt, &title.Seen, &title.SeenOn, &title.Poster, &title.ImdbRating, &title.ImdbID, &title.Year, &title.TomatoUserRating, &title.TomatoURL, &title.User, pq.Array(&title.Genres))
	if err != nil {
		http.Error(w, err.Error(), http.StatusNotFound)
		return
	}

	w.Header().Set("Content-Type", "application/json")
	json.NewEncoder(w).Encode(title)
}

func deleteTitle(w http.ResponseWriter, r *http.Request) {
	session, _ := store.Get(r, "session-name")
	userID := session.Values["user_id"].(string)

	vars := mux.Vars(r)
	titleID := vars["id"]

	query := `DELETE FROM titles WHERE id = $1 AND user = $2 RETURNING id, name, createdAt, seen, seenOn, poster, imdbRating, imdbID, year, tomatoUserRating, tomatoURL, user, genres`
	row := db.QueryRow(query, titleID, userID)

	var title Title
	err := row.Scan(&title.ID, &title.Name, &title.CreatedAt, &title.Seen, &title.SeenOn, &title.Poster, &title.ImdbRating, &title.ImdbID, &title.Year, &title.TomatoUserRating, &title.TomatoURL, &title.User, pq.Array(&title.Genres))
	if err != nil {
		http.Error(w, err.Error(), http.StatusNotFound)
		return
	}

	w.Header().Set("Content-Type", "application/json")
	json.NewEncoder(w).Encode(title)
}
